feature: @oneOf input object support

BobaFetters opened the pull request:
Adding support for @oneOf Input Objects, more info can be found in the official RFC here: https://github.com/graphql/graphql-spec/pull/825

Closes apollographql/apollo-ios#3430

BobaFetters committed:
Initial oneOf work

Added support for oneOf from graphql-js and new template renderer

BobaFetters committed:
Finished template starting tests for oneOf

BobaFetters committed:
OneOfInputObjectTemplate Unit Tests

BobaFetters committed:
Adding oneOf input to AnimalKingdomAPI

BobaFetters committed:
Merge branch 'main' into feature/oneof-3430

svc-apollo-docs commented:
### 🛠️ Docs Preview Building...

<details>
<summary>Configuration</summary>

```json
{
  "repoOverrides": {
    "apollographql/apollo-ios-dev@main": {
      "remote": {
        "owner": "apollographql",
        "repo": "apollo-ios-dev",
        "branch": "feature/oneof-3430"
      }
    }
  }
}
```

</details>

727629753b6f2aef0f9f3ebfbf6472f8a04138a2

BobaFetters commented on apollo-ios/Sources/ApolloAPI/SchemaTypes/InputObject.swift:
```diff
@@ -44,3 +44,5 @@ public struct InputDict: GraphQLOperationVariableValue, Hashable {
 
 }
+
+public protocol OneOfInputObject: InputObject { }
```
Went ahead and added this protocol stub to help differentiate the regular and oneOf input objects in the generated code, and also in case we need to add customizations or anything to oneOf objects specifically in the future.

AnthonyMDev replied:
I think this is a good idea.

AnthonyMDev commented on apollo-ios-codegen/Sources/ApolloCodegenLib/Templates/OneOfInputObjectTemplate.swift:
```diff
@@ -0,0 +1,53 @@
+  func renderBodyTemplate(
+    nonFatalErrorRecorder: ApolloCodegen.NonFatalError.Recorder
+  ) -> TemplateString {
+    let (validFields, deprecatedFields) = graphqlInputObject.fields.filterFields()
```
Getting warnings here for these properties not being used. Can we remove this line?

AnthonyMDev reviewed (approved):
Minor comment, but otherwise this looks good!

BobaFetters committed:
Removing unused fields

BobaFetters committed:
Merge branch 'main' into feature/oneof-3430

BobaFetters committed:
Updating Test Codegen config

Updating the SPMInXcodeProject codegen config to commit the animal kingdom package.swift and use local version of library

BobaFetters committed:
Revert "Updating Test Codegen config"

This reverts commit bb3cce46d493d5a1447eb16614fc382bbc9347ff.

calvincestari reviewed (approved):
Looks good Zach, thank you!

BobaFetters committed:
Merge branch 'main' into feature/oneof-3430

BobaFetters committed:
Updating codegen config test

BobaFetters committed:
Minor fixes

calvincestari commented on Tests/TestCodeGenConfigurations/SPMInXcodeProject/apollo-codegen-config.json:
```diff
@@ -57,7 +57,7 @@
       "moduleType" : {
         "swiftPackageManager" : {
-
+          
```
Should this one be changed to the new `swiftPackage` too?

BobaFetters replied:
Yea might as well, not actually required since this gets decoded into the new one but worth going ahead and switching it.

BobaFetters replied:
Actually looks like this one may require the change because of the need for the local path causing the ci failure

calvincestari commented:
Looks like there is still a CI test failure - is that CI related or more work to be done?

BobaFetters committed:
Update codegen config

BobaFetters committed:
Local path fix

BobaFetters committed:
Revert path change

BobaFetters committed:
Fixing spm generation issue

crsantos commented on Tests/TestCodeGenConfigurations/CodegenXCFramework/CodegenXCFramework/MyAPI/Schema/InputObjects/PetSearchInput.graphql.swift:
```diff
@@ -0,0 +1,23 @@
+import ApolloAPI
+
+public extension MyAPI {
+  enum PetSearchInput: OneOfInputObject {
```
Amazing that this fits like a glove on `enum`s with associated values ❤️ 

Thanks.
